web/client-api/src/main/java/io/deephaven/web/client/api/widget/plot/ChartData.java

```diff
@@ -29,118 +28,51 @@
-                assert nextAdded.getLast() < nextIndex;// the whole range should be there if any is
-
-                // update the index array and insert the actual data into our mapped columns
-                insertDataRange(tableData, nextAdded, i);
-
-                // increment i past these new items, so that our "next" is actually next
-                i += nextAdded.size();
-
-                nextAdded = addedIterator.hasNext() ? addedIterator.next() : null;
-            } else if (nextModified != null && nextModified.getFirst() == nextIndex) {
-                assert nextModified.getLast() >= nextIndex; // somehow being asked to update an item which wasn't
-                                                            // present
-
-                // the updated block is contiguous, make sure there are at least that many items to tweak
-                assert indexes.length - i >= nextModified.size();
-
-                replaceDataRange(tableData, nextModified, i);
-
-                // advance i past this section, since no other change can happen to these rows
-                i += nextModified.size();
-
-                nextModified = modifiedIterator.hasNext() ? modifiedIterator.next() : null;
-            } else if (nextRemoved != null && nextRemoved.getFirst() == nextIndex) {
-                assert nextRemoved.getLast() >= nextIndex; // somehow being asked to remove an item which wasn't present
-
-                // the block being removed is contiguous, so make sure there are at least that many and splice them out
-                assert indexes.length - i >= nextRemoved.size();
-
-                // splice out the indexes
-                asArray(indexes).splice(i, (int) nextRemoved.size());
-
-                // splice out the data itself
-                assert allColumns != null;
-                for (JsArray<Any> column : allColumns) {
-                    column.splice(i, (int) nextRemoved.size());
-                }
-
-                // don't in/decrement i, we'll just keep going
-                nextRemoved = removedIterator.hasNext() ? removedIterator.next() : null;
-            } else {
-
-                // no match, keep reading
-                i++;
+        while (removedPositions.hasNext()) {
+            Range nextRemoved = removedPositions.next();
+            for (JsArray<Any> column : allColumns) {
+                column.splice((int) nextRemoved.getFirst(), (int) nextRemoved.size());
             }
         }
 
-        if (JsSettings.isDevMode()) {
-            assert tableData.getRows().length == indexes.length;
-            assert cachedData.values().stream().flatMap(m -> m.values().stream())
-                    .allMatch(arr -> arr.length == indexes.length);
-            assert cachedData.values().stream().flatMap(m -> m.values().stream()).allMatch(arr -> arr
-                    .reduce((Object val, Any p1, int p2) -> ((Integer) val) + 1, 0) == indexes.length);
+        // then adds
```

Comment:
isn't this where you should apply shifts?

Reply:
(discussed on videocall) This implementation doesn't care about shifts - the rowSet provided by the subscription is already correctly shifted, we just need to splice the values into the (flat) array so that downstream charting can ingest those arrays directly.